rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,462 @@
+        return self._table.keys()
+
+    @property
+    def num_collisions_ignored(self) -> int:
+        """Returns the number of collisions that have been ignored."""
+        return self._num_collisions_ignored
+
+    @property
+    def num_collisions_resolved(self) -> int:
+        """Returns the number of collisions that have been resolved."""
+        return self._num_collisions_resolved
+
+    @classmethod
+    def _build_key(cls, sub_state: SubState) -> Tuple[Text, Text]:
+        """Builds the lookup table key for the given substate.
+
+        Expects the message to have exactly one of the  `KEY_ATTRIBUTES`.
+        Moreover, if `ENTITIES` is an attribute, then `TEXT` must be present.
+
+        Args:
+          sub_state: substate we want to store in the lookup table
+        Raises:
+          ValueErrors if there is more than one key attribute in the given substate
+          or if the substate contains `ENTITIES` but no `TEXT`
+        """
+        attributes = sub_state.keys()
+        key_attributes = set(attributes).intersection(cls.KEY_ATTRIBUTES)
+        if not key_attributes or len(key_attributes) > 1:
+            raise ValueError(
+                f"Expected exactly one attribute out of "
+                f"{cls.KEY_ATTRIBUTES} but received {attributes}"
+            )
+        if ENTITIES in attributes and TEXT not in attributes:
+            raise ValueError(
+                f"Expected entities information only in conjunction with `TEXT` "
+                f"but received a substate with {sub_state.keys()}."
+            )
+        key_attribute = list(key_attributes)[0]
+        return (key_attribute, str(sub_state[key_attribute]))
+
+    def add(self, message_with_one_key_attribute: Message) -> None:
+        """Adds the given message to the lookup table.
+
+        Args:
+          message_with_unique_key_attribute: The message we want to add to the lookup
+            table. It must have exactly one key attribute.
+
+        Raises:
+          ValueError if we cannot create a key for the given message or if this creates
+          a collision that we cannot resolve
+        """
+        key = self._build_key(sub_state=message_with_one_key_attribute.data)
+        existing_message = self._table.get(key)
+        if existing_message is not None:
+            if (
+                len(existing_message.features)
+                != len(message_with_one_key_attribute.features)
+                or existing_message.data.keys()
+                != message_with_one_key_attribute.data.keys()
+            ):
+                if self._handle_collisions:
+                    if len(existing_message.features) <= len(
+                        message_with_one_key_attribute.features
+                    ) and set(message_with_one_key_attribute.data.keys()) >= set(
+                        existing_message.data.keys()
+                    ):
+                        self._table[key] = message_with_one_key_attribute
```

Comment:
yep, think that's a good idea. And I might/could remove the extra collision handling stuff - if I manage to remove the entities from the lookup - I think I just messed something up last time because they should not be necessary (in the lookup table). And that means that we definitely don't need to resolve anything.

Reply:
that would be great as this would also reduce a lot of complexity / testing effort